File: haskell/p003.hs
-- Largest prime factor

factor :: Int -> Int -> Int
factor n i
    | n <= 1    = i
    | otherwise = factor (n `div` d) d where
        d = head [x | x <- [i..n], n `mod` x == 0]

largestPrimeFactor :: Int -> Int
largestPrimeFactor n = factor n 2


main :: IO ()
main = do
    print (largestPrimeFactor 600851475143)

File: java/p003.java
// Largest prime factor

public class p003 {
  long NUMBER = 600_851_475_143L;

  public static void main(String[] args) {
    System.out.println(new p003().solve());
  }

  public long solve() {
    long x = NUMBER;
    long i = 2;
    while (i < x) {
      if (x % i == 0) x /= i;
      else i++;
    }
    return i;
  }
}
